### PR TITLE
[Reland][Environment Variable][3/N] Use thread-safe getenv functions 

### PR DESCRIPTION
**Pull request by cyyever**
Reland of #137328, which was reverted due to reverting a dependent PR.

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

**Review by eqy — APPROVED** [review]

File: c10/cuda/CUDAMiscFunctions.cpp [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
 namespace c10::cuda {
 
+// NOLINTNEXTLINE(bugprone-exception-escape,-warnings-as-errors)
 const char* get_cuda_check_suffix() noexcept {
-  static char* device_blocking_flag = getenv("CUDA_LAUNCH_BLOCKING");
+  static auto device_blocking_flag =
+      c10::utils::check_env("CUDA_LAUNCH_BLOCKING");
   static bool blocking_enabled =
-      (device_blocking_flag && atoi(device_blocking_flag));
+      (device_blocking_flag.has_value() && device_blocking_flag.value());
```
Comment (ezyang) [inline c10/cuda/CUDAMiscFunctions.cpp:11]: This doesn't look semantics preserving

**Comment by ezyang** [comment]
Otherwise this one seems fine, ok not to revert

File: c10/cuda/CUDAMiscFunctions.cpp (hunk 1 above)
Comment (eqy) [reply c10/cuda/CUDAMiscFunctions.cpp:11]: +1 but do we think this is the likely culprit (seems rare to see anything outside of `CUDA_LAUNCH_BLOCKING=0` or `CUDA_LAUNCH_BLOCKING=1`

Comment (cyyever) [reply c10/cuda/CUDAMiscFunctions.cpp:11]: check_env assumes the value is "0" or "1"  and then coverts it to bool. So it's semantics preserving.

**Pull request closed without merging**